### ISSUE TITLE
all: rules for 3rd party packages

### ISSUE THREAD

**Issue by bradleyfalzon** [dominikh/go-tools#156]
It appears (haven't checked extensively) the checks to date have just focused on the Go language and standard library packages issues.

#155 appears to suggest staticcheck may also include rules for the `golang.org/x/` sub-repositories.

I couldn't find any definitive criteria for the rules. Is it just language, standard library and sub-repositories that are in scope?

I ask because there have been a few cases I've used 3rd party packages such as `github.com/pkg/errors` and there are simplifications that gosimple could suggest.

**Comment by mvdan** [comment]
Sorry, but where do you see `golang.org/x/` in there? `context` has been part of `std` since Go 1.7.

**Comment by dominikh** [comment]
tl;dr: we don't know yet.

There are no definitive criteria yet. So far we've only focused on the standard library; nobody asked for more than that, so we didn't have to make a decision. #155 would add the /x/ sub-repos. Here, we can turn a blind eye and pretend that it's "almost" the standard library.

For third party repositories I have toyed with the idea of supporting popular projects, but it opens a can of worms. What counts as popular, where do we draw the line, how do we communicate it, who is responsible for keeping checks up to date, how do we deal with packages changing their APIs but not providing versions we can check against?

Ideally, we'd open up the rule engine to external rule definitions, so that projects can provide their own rules, placed in an agreed-upon place. This idea is blocked on coming up with a good way of declaring such rules.

The "argument rule engine" we use in staticcheck was a step in this direction. It allows defining predicates for arguments (such as "argument 0 of function X must not be negative").  I am planning on eventually defining a format for defining such rules, possibly JSON, or a custom DSL.

But this doesn't help with more advanced checks, such as #155. Various products have various languages for pattern matching code, but none are satisfactory. I posit that even the approach used by `eg`, i.e. defining rules via plain Go syntax, will be insufficient to describe interesting defects, because you'd always want to add extra predicates that cannot be expressed in pure Go.

So the way I see it, we can allow projects to define simple API rules, but advanced checks depend on versioning to be solved.

**Comment by dominikh** [comment]
@mvdan What does `golang.org/x/sync/errgroup` have to do with `context`?

**Comment by mvdan** [comment]
Sorry, I hadn't seen `errgroup` :)

**Comment by bradleyfalzon** [comment]
Thanks, closing as that answers the question and others can still find this issue if they have the same query.
